### PR TITLE
[v11] backport #17870, #18021, #18286, part of #18299, and #19340

### PR DESCRIPTION
**Pull request by ibeckermayer**
Backports #17870, #18021, #18286, part of #18299, and #19340 to branch/v11

The primary intention of this backport is to backport #19340 to branch/v11. This is necessary, because the breaking Windows update that #19340 is preparing us for is scheduled for May 2023, at which point we expect to still be supporting Teleport 11. The other PRs included in this backport were those required to get everything to merge cleanly.

"Part of #18299" is just getting some of the changes from `lib/auth/windows/windows.go` that were made in #18299. This was done by running

```
git show 9c2192e -- lib/auth/windows/windows.go | git apply -
```

and then manually deleting `GenerateDatabaseCredentials` and `CertKeyPEM`.

### CODE REVIEW

**Comment by zmb3** [comment]
Do you need to backport all of the refactoring for SQL server?

Can we not just reinsert the SID mapping logic at the right place in branch/v11?

**Comment by ibeckermayer** [comment]
> Do you need to backport all of the refactoring for SQL server?
> 
> Can we not just reinsert the SID mapping logic at the right place in branch/v11?

It's of course possible to do this without the other refactoring, though it would be a lot of "surgery" to get right. Because of all the files and structs/funcs that have changed places, cherry-picking + manual conflict resolution is very convoluted. Probably it would be easiest to just copy paste things manually, without any rebasing.

However I think that this approach is preferable, primarily because it will make future backports simpler. If I do a bunch of manual surgery now, then I will need to continue doing it in the future if I make changes to the same things, and so on. IMO that risks adding more complexity to development over time than does backporting some of @alistanis' refactoring.

File: docs/pages/includes/desktop-access/desktop-config.yaml [hunk 1]
```diff
@@ -24,6 +24,19 @@ windows_desktop_service:
     # likely "EXAMPLE". When connecting as the "svc-teleport" user, you should
     # use the format: "EXAMPLE\svc-teleport".
     username: '$LDAP_USERNAME'
+    # The security identifier of the service account specified by the username
+    # field above. This looks like a string starting with "S-".
+    #
+    # An administrator with access to the domain controller can obtain this value
```
Comment (zmb3) [inline docs/pages/includes/desktop-access/desktop-config.yaml:30]: Do you actually need to be an administrator with access to a DC?

**Review by zmb3 — APPROVED** [review]

File: docs/pages/includes/desktop-access/desktop-config.yaml (hunk 1 above)
Comment (ibeckermayer) [reply docs/pages/includes/desktop-access/desktop-config.yaml:30]: From my research it seems like no. https://learn.microsoft.com/en-us/windows-server/identity/ad-ds/plan/security-best-practices/appendix-b--privileged-accounts-and-groups-in-active-directory#permissions. I believe any user besides a guest. I will change it in an upcoming PR to say

"Any AD user with the permission to read user objects can obtain this value..."

Comment (ibeckermayer) [reply docs/pages/includes/desktop-access/desktop-config.yaml:30]: Fixed in https://github.com/gravitational/teleport/pull/19914

**Review by rosstimothy — APPROVED** [review]

**Review by tigrato — APPROVED** [review]

**Pull request merged**